h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1404,69 +1407,76 @@ public void setMapId(int mapId) {
         }
 
         public Set<Integer> getReferenced() {
-            return referenced;
+            return referencedChunks;
         }
 
         public void visit(Page page) {
             long pos = page.getPos();
+            final int chunkId = DataUtils.getPageChunkId(pos);
             if (DataUtils.isPageSaved(pos)) {
-                register(DataUtils.getPageChunkId(pos));
+                registerChunk(chunkId);
             }
             int count = page.map.getChildPageCount(page);
-            if (count > 0) {
-                ChunkIdsCollector childCollector = getChild();
-                for (int i = 0; i < count; i++) {
-                    Page childPage = page.getChildPageIfLoaded(i);
-                    if (childPage != null) {
-                        childCollector.visit(childPage);
-                    } else {
-                        childCollector.visit(page.getChildPagePos(i));
-                    }
-                }
-                // and cache resulting set of chunk ids
-                if (DataUtils.isPageSaved(pos) && cacheChunkRef != null) {
-                    int[] chunkIds = childCollector.getChunkIds();
-                    cacheChunkRef.put(pos, chunkIds, Constants.MEMORY_ARRAY + 4 * chunkIds.length);
+            if (count == 0) {
+                return;
+            }
+            ChunkIdsCollector childCollector = getChild();
+            for (int i = 0; i < count; i++) {
+                Page childPage = page.getChildPageIfLoaded(i);
+                if (childPage != null) {
+                    childCollector.visit(childPage);
+                } else {
+                    childCollector.visit(page.getChildPagePos(i));
                 }
             }
+            // and cache resulting set of chunk ids
+            if (DataUtils.isPageSaved(pos) && cacheChunkRef != null) {
+                int[] chunkIds = childCollector.getChunkIds();
+                cacheChunkRef.put(chunkId, chunkIds, Constants.MEMORY_ARRAY + 4 * chunkIds.length);
+            }
         }
 
         public void visit(long pos) {
             if (!DataUtils.isPageSaved(pos)) {
                 return;
             }
-            register(DataUtils.getPageChunkId(pos));
-            if (DataUtils.getPageType(pos) != DataUtils.PAGE_TYPE_LEAF) {
-                int chunkIds[];
-                if (cacheChunkRef != null && (chunkIds = cacheChunkRef.get(pos)) != null) {
-                    // there is a cached set of chunk ids for this position
-                    for (int chunkId : chunkIds) {
-                        register(chunkId);
-                    }
+            final int chunkId = DataUtils.getPageChunkId(pos);
+            if (registerChunk(chunkId)) {
+                // if we have already visited this chunk
+                return;
```

Comment:
I think, shortcut logic introduced here is inherently faulty. Chunk is added to a set of referenced chunks, when it is discovered that it's still alive (have at list one live page in it). It does not mean that we've already look through all live pages that chunk may contain and all chunks referenced from them.